pkg/cluster/host.go

```diff
@@ -32,53 +35,91 @@ type Host interface {
 }
 
 type hostClient struct {
-	addr string
-	dial rpcplus.DialFunc
-	c    RPCClient
+	c *httpclient.Client
 }
 
 // NewHostClient creates a new Host that uses client to communicate with it.
-// addr and dial are used by Attach. dial may be nil to use the default dialer.
-func NewHostClient(addr string, client RPCClient, dial rpcplus.DialFunc) Host {
-	c := &hostClient{addr: addr, dial: dial, c: client}
-	if dial == nil {
-		c.dial = net.Dial
+// addr is used by Attach.
+func NewHostClient(addr string, h *http.Client) Host {
+	if h == nil {
+		h = http.DefaultClient
 	}
-	return c
+	return &hostClient{c: &httpclient.Client{
+		ErrPrefix:   "host",
+		ErrNotFound: ErrNotFound,
+		URL:         addr,
+		HTTP:        h,
+	}}
 }
 
 func (c *hostClient) ListJobs() (map[string]host.ActiveJob, error) {
 	var jobs map[string]host.ActiveJob
-	err := c.c.Call("Host.ListJobs", struct{}{}, &jobs)
+	err := c.c.Get("/host/jobs", &jobs)
 	return jobs, err
 }
 
 func (c *hostClient) GetJob(id string) (*host.ActiveJob, error) {
 	var res host.ActiveJob
-	err := c.c.Call("Host.GetJob", id, &res)
+	err := c.c.Get(fmt.Sprintf("/host/jobs/%s", id), &res)
 	return &res, err
 }
 
 func (c *hostClient) StopJob(id string) error {
-	return c.c.Call("Host.StopJob", id, &struct{}{})
+	return c.c.Delete(fmt.Sprintf("/host/jobs/%s", id))
 }
 
-func (c *hostClient) StreamEvents(id string, ch chan<- *host.Event) stream.Stream {
-	return rpcStream{c.c.StreamGo("Host.StreamEvents", id, ch)}
+func (c *hostClient) StreamEvents(id string, ch chan<- *host.Event) (Stream, error) {
+	header := http.Header{"Accept": []string{"text/event-stream"}}
+	r := fmt.Sprintf("/host/jobs/%s", id)
+	if id == "all" {
+		r = "/host/jobs"
+	}
+	res, err := c.c.RawReq("GET", r, header, nil, nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	stream := EventStream{
+		Chan: ch,
+		body: res.Body,
+	}
+	go func() {
+		defer func() {
+			close(ch)
+			stream.Close()
+		}()
+
+		r := bufio.NewReader(stream.body)
+		dec := sse.NewDecoder(r)
+		for {
+			event := &host.Event{}
+			if err := dec.Decode(event); err != nil {
+				if err != io.EOF {
```

Comment:
Pushed refactor to use reflection based approach as discussed on that tangent :D

Linecount went down nicely and I'm glad the error handling and disconnect handling is now uniform.